Add new feature resolver.

ehuss opened the pull request:
This adds a new resolver which handles feature unification independently of the main resolver. This can be enabled with the `-Zfeatures` flag which takes a comma-separated list of options to enable new behaviors. See `unstable.md` docs for details.

There are two significant behavior changes:

1. Ignore targets that are not enabled.
2. Do not unify features between build_deps, dev_deps, and normal deps.

The "forks" in the unit graph are handled by adding `DepKind` to `UnitFor`. The feature resolver tracks features independently for the different dependency kinds.

Unfortunately this currently does not support decoupling proc_macro dependencies. This is because at resolve time it does not know which dependencies are proc_macros. Moving feature resolution to after the packages are downloaded would require massive changes, and would make the unit computation much more complex. Nobody to my knowledge has requested this capability, presumably because proc_macros are relatively new, and they tend not to have very many dependencies, and those dependencies tend to be proc-macro specific (like syn/quote). I'd like to lean towards adding proc-macro to the index so that it can be known during resolve time, which would be much easier to implement, but with the downside of needing to add a new field to the index.

I did not update `cargo metadata`, yet. It's not really clear how it should behave. I think I'll need to investigate how people are currently using the feature information and figure out how it should work. Perhaps adding features to "dep_kinds" will be the solution, but I'm not sure.

The goal is to try to gather feedback about how well this new resolver works. There are two important things to check: whether it breaks a project, and how much does it increases compile time (since packages can be built multiple times with different features). I'd like to stabilize it one piece at a time assuming the disruption is not too great. If a project breaks or builds slower, the user can implement a backwards-compatible workaround of sprinkling additional features into `Cargo.toml` dependencies. I think `itarget` is a good candidate to try to stabilize first, since it is less likely to break things or change how things are built. If it does cause too much disruption, then I think we'll need to consider making it optional, enabled *somehow*.

There is an environment variable that can be set which forces Cargo to use the new feature resolver. This can be used in Cargo's own testsuite to explore which tests behave differently with the different features set.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #7731) made this pull request unmergeable. Please resolve the merge conflicts.

ehuss commented:
While looking through some of your questions, I found a bug with `cargo test` and dev-dependency handling. It might require some non-trivial changes to fix.  The problem is that in some cases the only difference between two `lib.rs` units is what it links against, and the current unit-deduplication logic can't handle that.  

I'll try to fix that, and then respond to your questions (and try to add some comments to clarify confusing sections).

ehuss commented:
So I think there may need to be significant changes, and I wanted to confirm that my assumptions make sense. I'll illustrate with an example project with these dependencies:

```toml
[dependencies]
dep = {version="1.0", features=["f1"]}

[dev-dependencies]
dep = {version="1.0", features=["f2"]}
```

With:
* `src/lib.rs`
* `src/main.rs`
* `tests/test.rs`
* `examples/example.rs`

`cargo build` will create a unit graph like the following:

![unit-graph-build](https://user-images.githubusercontent.com/43198/73111831-b9cbb700-3ec0-11ea-8b65-b746a2688009.png)

(Hopefully that's pretty straightforward.)

The tricky part comes with `cargo test`. I would assume with decoupled dev-dependencies, we would want a unit graph like the following:

![unit-graph-test](https://user-images.githubusercontent.com/43198/73111841-c4864c00-3ec0-11ea-9605-8bccffbcb756.png)

This causes `lib.rs` to be built 3 times: once as a test, once as a library linked with dev-dependencies (for other tests), and once as a library for examples and binaries (without dev-dependencies). This is done with the assumption that you want `main.rs` to be built the same as it does with `cargo build`.

If you agree with that, then the tricky bit comes with how to build that graph. The only difference for the `lib.rs` unit is what it links against. However, the `Unit` structure as it exists doesn't differentiate that. I'm thinking of adding `DepKind` to `Unit` to handle this properly. I'll try to avoid the extra build of `lib.rs` if you don't have any dev-dependencies, which will make it a little tricky (same with shared build-dependencies where there are no differences). I'm also wondering if I can change (or remove) `UnitFor`, which is essentially doing the same thing (the reason `UnitFor` exists is to encode these differences while still handling de-duplication). I'll try to put a fix together that does this, and see where it goes.

alexcrichton commented:
So that makes sense to me *if* we take into account `DepKind`. I'm not entirely convinced though (for now at least) that we should be doing that. Our current feature resolution is to unify based entirely on just the structure of the crate graph, but I'm not sure how much further we need to go other than unifying only based on target. For example I'm not sure if it's a bad thing where, in your example, `cargo build` works the same but `cargo test` only builds `dep` with both features enabled.

ehuss commented:
Yea, I'm not really clear on what the best behavior here is. How about this:  have some sort of internal boolean (`CompileFilter::need_dev_deps`) that indicates whether or not dev-dependency features should be included.  It would be enabled if any test/bench/example is being built.  That means that `cargo test` would behave the same as it does today.  It would also mean something like `cargo build --all-targets` would also unify (the same as it does today).  But if you do a plain `cargo build`, it will build without unifying dev-dependencies.

I'm not sure if that might be too confusing or surprising.  Perhaps with documentation it can be made clear?

I'm inclined to go that direction because it will be much simpler to implement.  If we decide that dependencies should be "forked" and built multiple times with different features, we can add that in the future.

How does that sound?

alexcrichton commented:
That sounds relatively reasonable to me yeah, the difference in `--all-targets` behavior is a little unfortunate but I think that basically mirrors what happens today (albeit probably in a slightly different place)

ehuss commented on src/cargo/core/compiler/unit_dependencies.rs:
```diff
@@ -271,7 +286,16 @@ fn compute_deps<'a, 'cfg>(
+            DepKind::Development
+        } else {
+            DepKind::Normal
+        };
```
> First and foremost, this seems like it's "the" spot that the magic happens in terms of differentiating what kind of features go where, is that right?

Yes.

> If our target `is_custom_build`, that means our target is the execution (?) of a build script,

The building, not the execution.  Execution units (`RunCustomBuild`) are created in `dep_build_script`, and computing the *running* deps is done in `compute_deps_custom_build`.

But, yes, build scripts only have build dependencies.

> To confirm, the desired behavior here is…  `log` is recompiled in test mode, right?

Correct.  

Per our previous conversation, I have changed things so that `-Zfeatures=dev_dep` is essentially disabled if any test, bench, or example is being built.

> So all in all I'm a little lost about why this `any` isn't an `all` for the development dep kind. Could you help clear that up for me?

Why would it be `all`?  In the case of `cargo build`, the dev-dependencies are already filtered out (a few lines above).  In the case of `cargo test`, in your example with `log` defined twice, the tests will see two deps (one normal, one dev), and the test units want to use dev-dependency features, so it needs to be `any`.

bors commented:
:umbrella: The latest upstream changes (presumably #7851) made this pull request unmergeable. Please resolve the merge conflicts.

ehuss commented on src/cargo/core/compiler/unit_dependencies.rs:
```diff
@@ -353,6 +377,7 @@ fn compute_deps<'a, 'cfg>(
 /// the returned set of units must all be run before `unit` is run.
 fn compute_deps_custom_build<'a, 'cfg>(
     unit: &Unit<'a>,
+    dep_kind: DepKind,
```
For the purpose of computing features, building the `build.rs` needs to use the features from the parent package.  If it always used "build", it would pick the wrong features for shared dependencies.  Example:

```
foo
├── foo build.rs (RunCustomBuild) NORMAL features
│   └── foo build.rs (Build) NORMAL features
│       └── common 
│           └── common build.rs (RunCustomBuild) BUILD-DEPENDENCY features
│               └── common build.rs (Build) BUILD-DEPENDENCY features
└── common
    └── common build.rs (RunCustomBuild) NORMAL features
        └── common build.rs (Build) NORMAL features
``` 

Here, when creating the `common build.rs` units, it wants the features based on its context (within a build dependency or not).  This is important for building build.rs because scripts often use cfg!() expressions for determining which features are enabled (as opposed to environment variables).

ehuss commented on src/cargo/core/compiler/unit_dependencies.rs:
```diff
@@ -486,6 +512,7 @@ fn maybe_lib<'a>(
 /// build script.
 fn dep_build_script<'a>(
     unit: &Unit<'a>,
+    dep_kind: DepKind,
```
Similar to above, running the build script needs to set the `CARGO_FEATURE_*` environment variables for its parent package.  It needs to know if the parent is being built as a build-dependency or not so it can set those env vars correctly.

ehuss commented on src/cargo/core/profiles.rs:
```diff
@@ -869,23 +886,33 @@ impl UnitFor {
         static ALL: &[UnitFor] = &[
             UnitFor {
                 build: false,
+                dep_kind: DepKind::Normal,
```
I have rewritten to drop `dep_kind` in favor of a `build_dep` bool.  Hopefully that will be a little more straightforward.

ehuss commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,549 @@
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::rc::Rc;
+
+type ActivateMap = HashMap<(PackageId, DepKind, CompileKind), BTreeSet<InternedString>>;
```
Yep, added a comment.  Also changed it, as CompileKind and DepKind wasn't really the right distinction.

ehuss commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -720,7 +738,12 @@ fn generate_targets<'a>(
+            pkg.package_id(),
+            DepKind::Normal,
+            kind,
+        ));
```
Root units are not dependencies of anything. Their features come from the command-line options.

For test targets, nothing can depend on a test, so there are no edges going from dev-dependency to the test. Thus there shouldn't be any dev-dependency features injected into the test.

proc_macros are not supported (because we don't know they exist at resolve time).

I've been trying to think if there's some way to use `[patch]` or cyclical dev-dependencies to break these assumptions, but I haven't been able to come up with anything.

ehuss commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,549 @@
+    requested_target: CompileKind,
+    resolve: &'a Resolve,
+    /// Features requested by the user on the command-line.
+    requested_features: &'a RequestedFeatures,
```
Agreed, moved to parameters instead.

ehuss commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,549 @@
+    fn do_resolve(&mut self) -> CargoResult<()> {
+        if self.opts.package_features {
+            let mut found = false;
+            for member in self.ws.members() {
```
Moved out to a common function in `Workspace`.

ehuss commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,549 @@
+        pkg_id: PackageId,
+        requested_features: &RequestedFeatures,
+    ) -> CargoResult<()> {
+        let fvs = self.fvs_from_requested(pkg_id, CompileKind::Host, requested_features);
```
This no longer tracks DepKind or CompileKind.

ehuss commented:
I did a fairly major rewrite to try to simplify things. Now that dev-dependency unification is a global setting, it all really boils down to whether or not something is a build dependency, so I just made that a bool.  It's not as flexible, but I can always switch it back to something more complex if needed.

I also add `-Zfeatures=all` to try all options at once.

BTW, I'd like to squash this before merging, so don't merge as-is.  This also now includes #7857 to fix some issues, but that PR should probably go first to simplify this.

ehuss commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -897,15 +929,22 @@ fn generate_targets<'a>(
 
 fn resolve_all_features(
```
I fixed a bug where required-features was not handling build-dependencies correctly.  See also #7855 and #7853 for some other bugs in this function.

shamatar commented:
Does this PR also address #2589 and #2524 ? If not I'd try to look deeper into the 2589 myself as it would be quite convenient to have something like

```
[target.'cfg(not(target_os = "macos"))'.dependencies.a]
version = "0.5"

[target.'cfg(target_os = "macos")'.dependencies.a]
git = "https://github.com/foo/a_patched.git"
```

ehuss commented:
@shamatar yes, it should close those issues (and many others).  2589 relates to features, not multiple sources.  Multiple sources is #7753, and I believe that will be very difficult to fix.

shamatar commented:
@ehuss 

Ok, thanks for an answer. Another short question: if paths are not easy to fix, may be it would be able to introduce a syntax like
```
[target.'cfg(not(target_os = "macos"))'.features]
default = []

[target.'cfg(target_os = "macos")'.features]
default = ["macos_patch"]
```

or something similar. If it ever makes sense. At least it's a short-term fix for different paths problem.

ehuss commented:
That is discussed in #1197.

bors commented:
:umbrella: The latest upstream changes (presumably #7855) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
Ok this is taking me longer to get to reviewing than I thought it might. I'll be gone most of next week as well. From my previous review and given the changes you've described implementing, I'd be fine landing this in the meantime and I can comment on this when I get a chance to. Or if you're ok waiting a week or so that's ok too!

matklad commented:
>I did not update cargo metadata, yet. It's not really clear how it should behave. I think I'll need to investigate how people are currently using the feature information and figure out how it should work.

I guess I can say a thing or two about IntelliJ and rust-analyzer, who are perhaps the two biggest users of `cargo metadata` at the moment. In short, the current `Resolve` in `cargo metadata` doesn't really make sense; it works good enough, but it is on the wrong level of abstraction. 

Specifically, Resolve is on the level of packages, but what rust-analyzer & IntelliJ want is more close to units. For example, we would like to see "crate ./test/foo.rs depends on crate ./src/lib.rs with these cfg flags set". 

I think we can make two steps to make sure that things "make sense".

1. Recast existing Resolve as an exact reflection of `Cargo.lock`. Ie, make it fully independent of the set of feature flags and the target, such that it is basically the flat list of packages from a lockfile + a deps which explain *why* those packages are in lockfile. I am not sure if this would be actually useful, but at least that would be a clear defined things

2. Add a separate thing (perhaps a new command even?) which outputs a format close to the UnitMap, which speaks not about Cargo packages, but rather crates in rust-reference sense, and dependencies between them. This would depend on the precise set of flags passed to Cargo (we'll get different outptus for `cargo build` vs `cargo test`) and would actually resemble a sort-of static build plan.

ehuss commented:
@matklad I've been thinking the exact same thing.  I've also been wondering if we could just extend `cargo metadata` or need some other command or flag (like how `--build-plan` works).

I'd really like a tool to visualize the unit graph, but that is difficult for any non-trivial project.

sunshowers commented:
FYI cargo-guppy relies on metadata as well: https://github.com/calibra/cargo-guppy

alexcrichton commented on src/cargo/core/compiler/standard_lib.rs:
```diff
@@ -99,8 +102,13 @@ pub fn resolve_std<'cfg>(
-    let resolve = ops::resolve_ws_with_opts(&std_ws, opts, &specs)?;
-    Ok((resolve.pkg_set, resolve.targeted_resolve))
+    let resolve =
+        ops::resolve_ws_with_opts(&std_ws, target_data, requested_target, &opts, &specs, false)?;
```
Reading up on this, the `false` here seems pretty special, so could it perhaps transition to an enum with named stuff? It's definitely more verbose, but we probably want to err on the side of making sure this is as resilient as can be!

alexcrichton commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,504 @@
+
+impl ResolvedFeatures {
+    /// Returns the list of features that are enabled for the given package.
+    pub fn activated_features(&self, pkg_id: PackageId, is_build: bool) -> Vec<InternedString> {
```
Similar to a different comment elsewhere, could `is_build` be replaced with something more desciptive? I'm imagining something like two methods (`activated_features` and `activated_features_for_build_dep`) or perhaps an `enum` with named variants here.

alexcrichton commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,504 @@
+                            self.activate_pkg(
+                                dep_pkg_id,
+                                &fvs,
+                                for_build || (self.opts.decouple_build_deps && dep.is_build()),
```
FWIW this clause is showing up a number of times, could this boolean expression's result be stored in the vector returned by `self.deps`? I think it's cacheable there, right?

alexcrichton commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,504 @@
+                // Activate the optional dep.
+                for (dep_pkg_id, deps) in self.deps(pkg_id, for_build) {
+                    for dep in deps {
+                        if dep.name_in_toml() == *dep_name {
```
I'm personally a fan of reducing rightward-drift wherever possible so if you're also a fan feel free to switch this to a `continue` followed by de-indenting the body

alexcrichton commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,504 @@
+        for fv in fvs {
+            self.activate_fv(pkg_id, fv, for_build)?;
+        }
+        if !self.processed_deps.insert((pkg_id, for_build)) {
```
This guard seems a bit suspicious to me. Just because we visited a package before here doesn't mean we didn't later add features which themselves add dependencies, right?

alexcrichton replied:
I'm thinking of something like `cargo build -p deep-dependency -p root-dependency` where it activates `deep-dependency` quickly but by activating `root-dependency` we transitively enable more features of `deep-dependency`

ehuss replied:
I think it is fine.  After resolving deep-dependency once, other packages that add features to that deep-dependency will do so via `FeatureValue::CrateFeature` in `activate_fv`.  This in turn checks if it is activating an optional dependency, and will recurse into that if necessary.

This is kinda fundamental to the way this is organized.  If you notice, a few lines below it *always* skips optional dependencies. Whenever a dependency is encountered that enables an optional feature, it will enable it and recurse right away. 

I believe this is structured somewhat differently from how the dependency resolver works.  Presumably because it is figuring out the features as it goes along, and doesn't have the full graph built, yet.

alexcrichton replied:
Ah ok the skipping of optional deps below actually does make this sense yeah, sorry I missed that! It still feels a bit funky to me but because it's doing the skip I'm fine deferring to tests to verify the correctness of this. Mind adding a comment here clarifying what sort of recursion this is guarding against, and how more recursion is actually allowed via feature activation below in a limited way?

alexcrichton commented on src/cargo/core/profiles.rs:
```diff
@@ -766,7 +766,17 @@ pub struct UnitFor {
+    /// features of the non-host package (whereas `host` is true because the
+    /// build script is being built for the host). `build_dep` becomes `true`
+    /// for build-dependencies, or any of their dependencies.
+    build_dep: bool,
```
I'm thinking about this for a bit now. I understand (I think?) all the words in this comment, but I'm having trouble piecing together a situation about where this becomes relevant. Can you give a small example about where this subtle distinction is necessary?

Additionally, since `host=false, build_dep=true` probably doesn't make much sense, would it make more sense for this to be some sort of tri-value thing like:

```rust
enum Kind {
    Target,
    Host { build_dep: bool },
}
```

alexcrichton replied:
(please don't use `Kind` as I'm sure you're already thinking when reading that)

ehuss replied:
I added an example.

I'm reluctant to add an enum to represent the 3 states. All the code is oriented around simple booleans, and I think it would make it harder to follow if they were unified into one field. I usually agree that it is good to prevent impossible scenarios, but in this case I think it might be harder to understand.

alexcrichton replied:
Ok the example looks good to me, thanks! I'm also fine deferring to your thoughts about booleans! 

One final question here, though. Technically we have two sets of fetures to deal with, one when the build script is compiled (the `--feature` flags) and one when it's executed (`CARGO_FEATURE_*`). Given that `is_custom_build` is conflating both executing and compiling the build script, I think we may want to be a little more nuanced/precise? I would expect that the compilation of the build script would use the `host: true` and `build_dep: false` set of features, but the execution of the build script would use `host: false` and `build_dep: false` as well, right? (ok maybe this doesn't have to do with `build_dep`?)

Also to clarify, in the example you listed here, `shared_dep build.rs` is only compiled once regardless of feature settings, right? And then it's executed twice maybe depending on feature settings/targets?

ehuss replied:
Hm, so it is working correctly, but you do point out that `host` isn't accurate for the RunCustomBuild unit.  `host` only matters for the profile, and the RunCustomBuild unit goes through a separate code path for computing the profile.  That is [here](https://github.com/rust-lang/cargo/blob/369991b9a09c9f49deebe81fa05d6e104c59ea11/src/cargo/core/compiler/unit_dependencies.rs#L496-L501) where it manually fetches the profile (ignoring unit_for).   All other units go through [this path](https://github.com/rust-lang/cargo/blob/369991b9a09c9f49deebe81fa05d6e104c59ea11/src/cargo/core/compiler/unit_dependencies.rs#L544-L549) where it inspects unit_for to determine the profile.  However, it is necessary for `host` to be `true` for RunCustomBuild so that all dependencies below it are marked as "host:true".  Added some comments explaining this.

> `shared_dep build.rs` is only compiled once regardless of feature settings, right? And then it's executed twice maybe depending on feature settings/targets?

It depends.  With `-Zfeatures=build_dep`, and the features listed are different between the normal and build dep, then it will get built twice.  The reason for this is that many scripts out there use `#[cfg]` and `cfg!` to detect features, instead of inspecting "CARGO_FEATURE_*".   Only building it once would break a large number of build scripts.  I don't think it would be feasible to avoid that.

alexcrichton replied:
Ah ok that makes sense, I know I've "used and abused" that `CARGO_FEATURE_*` is the "same" as `#[cfg]` in the past myself, so having a special case for this makes sense.

alexcrichton commented on src/cargo/core/compiler/unit_dependencies.rs:
```diff
@@ -168,7 +178,7 @@ fn deps_of_roots<'a, 'cfg>(roots: &[Unit<'a>], mut state: &mut State<'a, 'cfg>)
                 // This normally doesn't happen, except `clean` aggressively
                 // generates all units.
-                UnitFor::new_build()
+                UnitFor::new_build(false)
```
To make sure I understand, this is where the "subtelty" of `build_dep` happens, right? I depend on the `host` version of a build script being compiled, but the compilation doesn't have the `build_dep` flag to `true`? (but transitive deps have it set to `true`).

I may also be getting mixed up again about what `is_custom_build` is again and which is "run the script" and which is "build the script"

ehuss replied:
`is_custom_build` is true for both `CompileMode::RunCustomBuild` and `CompileMode::Build` of a build.rs script.  Normally code uses `is_run_custom_build` if it needs just one or the other.

Yea, any situation there is `UnitFor::new_build(false)` is the "subtle" difference.  That is, whenever adding a `build.rs` script for a normal dependency.

alexcrichton replied:
Ok sounds good to me!

alexcrichton commented on src/cargo/core/profiles.rs:
```diff
@@ -858,36 +878,65 @@ impl UnitFor {
+    ///
+    /// This is part of the machinery responsible for handling feature
+    /// decoupling for build dependencies in the new feature resolver.
+    pub fn with_build_dep(mut self, build_dep: bool) -> UnitFor {
```
As a minor nuance (and this was already present, just something I'm realizing now) the terminology of `with_*` doesn't quite fit here in my mind since I'd expect a function of that name to unconditionally configure `build_dep` with the argument provided, but here it's more of a unioning function. I'm not sure of a better name though!

ehuss replied:
Yea, I struggled a bit with the naming convention and wasn't happy with it.  But I'm unable to think of anything particularly better.  If you think of something, let me know!

I don't know if there is a common naming convention for this idiom (creating a copy and modifying a field in one step)?  Maybe that's a sign it may be better to be explicit about creating a copy, and then call `set_build_dep`?

alexcrichton replied:
Eh it's a minor thing, I wouldn't stress out much over it. I definitely know of no naming convention for this, so it may be good to just highlight it in the documentation and move on. (the current call-sites all "look right" which is what matters)

ehuss commented:
Added some comments, and fixed a critical bug in how shared optional dependencies were handled.

@Eh2406 you mentioned in the meeting wanting a high level overview, but I wasn't sure if you meant beyond what is already written.  Is there additional information you'd like to see in this PR?  I think if you start by reading `unstable.md` to get a sense of what it does and how to use it, you can then go to `resolve/features.rs` which contains an overview of how it fits.  I've tried to include plenty of comments, but let me know if there's anything you'd like me to add.

alexcrichton commented:
Alright I'm definitely happy with everything here, so I'm going to r+

I think it's safe to say this has languished long enough (sorry!) and we can continue to iterate in-tree. Thanks again so much for pushing on this @ehuss!


@bors: r+

bors commented:
:pushpin: Commit 4d0fda7cd9efb14055915e755b09ee3cf64586ff has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 4d0fda7cd9efb14055915e755b09ee3cf64586ff -->
<!-- homu: {"type":"Approved","sha":"4d0fda7cd9efb14055915e755b09ee3cf64586ff","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit 4d0fda7cd9efb14055915e755b09ee3cf64586ff with merge d6fa260d1aa74521c7d8c03d3d3c09067a4147d0...
<!-- homu: {"type":"BuildStarted","head_sha":"4d0fda7cd9efb14055915e755b09ee3cf64586ff","merge_sha":"d6fa260d1aa74521c7d8c03d3d3c09067a4147d0"} -->

Eh2406 commented on src/cargo/core/resolver/features.rs:
```diff
@@ -0,0 +1,534 @@
+//!
+//! This is a new feature resolver that runs independently of the main
+//! dependency resolver. It is intended to make it easier to experiment with
+//! new behaviors. When `-Zfeatures` is not used, it will fall back to using
```
This would be a good place to explain that it is not going to replace the dependency resolver but that it acts as a second pass. By not considering changing which versions are selected it avoids solving the NP-Hard problem faced by the dependency resolver. And it is focused on what should be built for this system not `all possible systems` that the dependency resolver considers. It therefore has complexity budget to consider ... (the list of really cool things your system achieves.)

ehuss replied:
Posted #7918 if you could review at your leisure.

> And it is focused on what should be built for this system not `all possible systems` that the dependency resolver considers.

This part I didn't quite follow.  The dependency resolver runs twice, once with all features enabled (which I think is what you mean by "all possible systems"?), and again with just the features on the command-line.  The new feature resolver has the same capabilities, but it only runs once.  By "same", I mean if you do `--all-features --workspace`, then that should result in the same "all possible" scenario matching `Cargo.lock`'s output.

Eh2406 replied:
I think I meant a case like:
```
a depends on b and c
b depends on e with feature eb
c depends on e only on MacOS e with feature ec 
e exists
```
What features does `e` have when built on a Windows system? `dependency resolver` thinks `{eb, ec}`. I think, the new features resolver can see that no active dependency asked for `ec`. Is that true?

ehuss replied:
Correct, `ec` is not activated.

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=21412)
Approved by: alexcrichton
Pushing d6fa260d1aa74521c7d8c03d3d3c09067a4147d0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=21412"},"merge_sha":"d6fa260d1aa74521c7d8c03d3d3c09067a4147d0"} -->

cbeck88 commented:
Thank you so so much!!!
